src/policy/policy.cpp

```diff
@@ -154,6 +154,97 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
+
+            std::vector <std::vector<unsigned char>> vSolutions;
+            txnouttype whichType;
+            // get the scriptPubKey corresponding to this input:
+            CScript prevScript = prev.scriptPubKey;
+            Solver(prevScript, whichType, vSolutions);
+
+            if (whichType == TX_SCRIPTHASH) {
+                std::vector <std::vector<unsigned char>> stack;
+                // If the scriptPubKey is P2SH, we try to extract the redeemScript casually by converting the scriptSig
+                // into a stack. We do not check IsPushOnly nor compare the hash as these will be done later anyway.
+                // If the check fails at this stage, we know that this txid must be a bad one.
+                if (!EvalScript(stack, tx.vin[i].scriptSig, SCRIPT_VERIFY_NONE, BaseSignatureChecker(), SIGVERSION_BASE))
+                    return TX_BAD_P2SH;
+                if (stack.empty())
+                    return TX_BAD_P2SH;
+                prevScript = CScript(stack.back().begin(), stack.back().end());
+            }
+
+            int witnessversion = 0;
+            std::vector<unsigned char> witnessprogram;
+
+            // Non-witness program must not be associated with any witness
+            if (!prevScript.IsWitnessProgram(witnessversion, witnessprogram) && !tx.wit.vtxinwit[i].IsNull())
+                return TX_BAD_WITNESS;
+
+            // Witness for P2WPKH must have 2 stack items. Pubkey must match the witness program.
+            // Signature must be <= 73 bytes (BIP66). Since signature size is variable, it is not impossible to bloat a
+            // 70-byte signature to 73-byte and trigger fee rejection for marginal cases but the impact is negligible.
+            if (witnessversion == 0 && witnessprogram.size() == 20) {
+                if (tx.wit.vtxinwit[i].scriptWitness.stack.size() != 2)
+                    return TX_BAD_WITNESS;
+                std::vector<unsigned char> hashPubKey(20);
+                std::vector<unsigned char> pubKey = tx.wit.vtxinwit[i].scriptWitness.stack[1];
+                if (tx.wit.vtxinwit[i].scriptWitness.stack[0].size() > 73)
+                    return TX_BAD_WITNESS;
+                CHash160().Write(begin_ptr(pubKey), pubKey.size()).Finalize(begin_ptr(hashPubKey));
+                if (hashPubKey != witnessprogram)
+                    return TX_BAD_WITNESS;
+            }
+
+            // P2WSH requires at least 1 witness stack item but we already know the witness is not empty.
+            // Make sure the witnessScript size is <= 10000 bytes and matches witness program.
+            if (witnessversion == 0 && witnessprogram.size() == 32) {
+                CScript witnessScript = CScript(tx.wit.vtxinwit[i].scriptWitness.stack.back().begin(), tx.wit.vtxinwit[i].scriptWitness.stack.back().end());
+                uint256 hashWitnessScript;
+                if (witnessScript.size() > 10000)
```

Comment:
Please use the constant found here: https://github.com/bitcoin/bitcoin/blob/master/src/script/script.h#L31